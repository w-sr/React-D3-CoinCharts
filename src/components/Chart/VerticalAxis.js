import React from "react";
import currencyFormatter from "currency-formatter";
import { extent } from "d3-array";
import styled from "styled-components";

import Flex from "../Flex";
import { color, fontSize, fontWeight } from "../../styles/constants";
import { PROPTYPES } from "../../constants";

const StyledVerticalAxis = styled(Flex).attrs({
  alignItems: props => (props.align === "right" ? "flex-end" : "flex-start"),
})`
  background-color: transparent;
  align-items: ${props => props.alignItems};
  flex: 0 0 3em;
  max-width: 3em;
  padding: 0.5em 0;
  z-index: 10;
`;

const Tick = styled(Flex)`
  color: ${color.coinchartsGray};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};
  text-align: center;
`;

const VerticalAxis = ({ data, align, currency }) => {
  const formatPrice = price => currencyFormatter.format(price, { code: currency, precision: 0 });
  const [minPrice, maxPrice] = extent(data, d => d.price);

  return (
    <StyledVerticalAxis data-testid="VerticalAxis" column justify="space-between" align={align}>
      {maxPrice && <Tick>{formatPrice(maxPrice)}</Tick>}
      {minPrice && <Tick>{formatPrice(minPrice)}</Tick>}
    </StyledVerticalAxis>
  );
};

VerticalAxis.propTypes = {
  currency: PROPTYPES.CURRENCY.isRequired,
  data: PROPTYPES.PRICE_DATA.isRequired,
  align: PROPTYPES.ALIGNMENT.isRequired,
};

export default VerticalAxis;
