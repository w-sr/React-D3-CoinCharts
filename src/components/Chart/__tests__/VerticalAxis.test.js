import React from "react";
import { render } from "react-testing-library";

import VerticalAxis from "../VerticalAxis";

const MAX_PRICE = 999;
const MIN_PRICE = 1;
const TEST_PROPS = {
  currency: "USD",
  data: [
    { price: 10, time: new Date(1000000) },
    { price: 20, time: new Date(2000000) },
    { price: 30, time: new Date(3000000) },
    { price: 40, time: new Date(4000000) },
    { price: 50, time: new Date(5000000) },
    { price: 60, time: new Date(6000000) },
    { price: MAX_PRICE, time: new Date(7000000) },
    { price: MIN_PRICE, time: new Date(8000000) },
  ],
  align: "left",
};

describe("<VerticalAxis />", () => {
  it("renders without crashing", () => {
    const { container } = render(<VerticalAxis {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders ticks correctly", () => {
    const { container } = render(<VerticalAxis {...TEST_PROPS} />);
    expect(container.firstChild.children).toHaveLength(2);
    expect(container.firstChild.children[0].textContent).toEqual(`$${MAX_PRICE}`);
    expect(container.firstChild.children[1].textContent).toEqual(`$${MIN_PRICE}`);
  });

  it("does not render ticks when `props.data` is []", () => {
    const { container } = render(<VerticalAxis {...TEST_PROPS} data={[]} />);
    expect(container.firstChild.children).toHaveLength(0);
    expect(container.firstChild).toMatchSnapshot();
  });
});
