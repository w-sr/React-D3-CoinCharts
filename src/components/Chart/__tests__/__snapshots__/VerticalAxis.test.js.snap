// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerticalAxis /> does not render ticks when \`props.data\` is [] 1`] = `
<div
  class="sc-bwzfXH hYXZTS sc-bdVaJa ivjahH"
  data-testid="VerticalAxis"
/>
`;

exports[`<VerticalAxis /> renders without crashing 1`] = `
<div
  class="sc-bwzfXH hYXZTS sc-bdVaJa ivjahH"
  data-testid="VerticalAxis"
>
  <div
    class="sc-htpNat eZyXfs sc-bdVaJa cAcgKX"
  >
    $999
  </div>
  <div
    class="sc-htpNat eZyXfs sc-bdVaJa cAcgKX"
  >
    $1
  </div>
</div>
`;
